ksql-engine/src/test/java/io/confluent/ksql/topic/DefaultTopicInjectorTest.java

```diff
@@ -0,0 +1,288 @@
+
+      // When:
+      fixture.inject();
+    }
+
+    @Test
+    public void testMalformedReplicaPropertyInOverrides() {
+      // Given:
+      fixture.givenStatement(
+          "CREATE STREAM sink WITH (value_format='JSON') AS SELECT * FROM source;"
+      );
+      fixture.propertyOverrides.put(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, "hi");
+
+      // Expect:
+      expectedException.expect(KsqlException.class);
+      expectedException.expectMessage(
+          "Invalid property override ksql.sink.replicas: hi "
+              + "(all overrides: {ksql.sink.replicas=hi})");
+
+      // When:
+      fixture.inject();
+    }
+
+    @Test
+    public void testMalformedPartitionsProperty() {
+      // Given:
+      fixture.givenStatement(
+          "CREATE STREAM sink WITH (partitions='hi', value_format='JSON') AS SELECT * FROM source;"
+      );
+
+      // Expect:
+      expectedException.expect(KsqlException.class);
+      expectedException.expectMessage("Invalid number of partitions in WITH clause: 'hi'");
+
+      // When:
+      fixture.inject();
+    }
+
+    @Test
+    public void testMalformedPartitionsPropertyNumber() {
+      // Given:
+      fixture.givenStatement(
+          "CREATE STREAM sink WITH (partitions=.5, value_format='JSON') AS SELECT * FROM source;"
+      );
+
+      // Expect:
+      expectedException.expect(KsqlException.class);
+      expectedException.expectMessage("Invalid number of partitions in WITH clause: 0.5");
+
+      // When:
+      fixture.inject();
+    }
+
+    @Test
+    public void testMalformedPartitionsPropertyInOverrides() {
+      // Given:
+      fixture.givenStatement(
+          "CREATE STREAM sink WITH (value_format='JSON') AS SELECT * FROM source;"
+      );
+      fixture.propertyOverrides.put(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY, "hi");
+
+      // Expect:
+      expectedException.expect(KsqlException.class);
+      expectedException.expectMessage(
+          "Invalid property override ksql.sink.partitions: hi "
+              + "(all overrides: {ksql.sink.partitions=hi})");
+
+      // When:
+      fixture.inject();
+    }
+
+    @Test
+    public void shouldDoNothingForNonCAS() {
+      // Given:
+      final PreparedStatement<?> statement = fixture.givenStatement(
+          "SHOW TOPICS;"
+      );
+
+      // When:
+      final PreparedStatement<?> injected = fixture.inject(statement);
+
+      // Then:
+      assertThat(injected, is(sameInstance(statement)));
+    }
+  }
+
+  /**
+   * This class tests all combinations of possible ways to get the topic
+   * information. There are four place to get the information: with clause,
+   * overrides, the ksql config and the source. For each place, there are
+   * four options: both partitions/replicas are present, only partitions is
+   * present, only replicas is present and nothing is present.
+   */
+  @RunWith(Parameterized.class)
+  public static class PrecedenceTest {
```

Comment:
Personally, I prefer the format I suggested for defining the precedence in the tests themselves..  This test is very hard to read.  In fact I'm not even going to try.

Reply:
I will unroll the loop. But there are 64 possible combinations and writing a different test for each one of those is super cumbersome.